R/lazy-join-query.R

```diff
@@ -1,34 +1,92 @@
 #' @export
 #' @rdname sql_build
-lazy_join_query <- function(x, y, vars, type = "inner", by = NULL, suffix = c(".x", ".y"), na_matches = FALSE) {
-  structure(
-    list(
-      x = x,
-      y = y,
-      vars = vars,
-      type = type,
-      by = by,
-      suffix = suffix,
-      na_matches = na_matches,
-      last_op = "join"
-    ),
-    class = c("lazy_join_query", "lazy_query")
+lazy_join_query <- function(x,
+                            y,
+                            vars,
+                            type,
+                            by,
+                            suffix = c(".x", ".y"),
+                            na_matches = c("never", "na")) {
+  stopifnot(inherits(x, "lazy_query"))
+  stopifnot(inherits(y, "lazy_query"))
+  join_check_vars(vars)
+  type <- arg_match(type, c("left", "right", "inner", "full", "cross"))
+  join_check_by(by)
+  vctrs::vec_assert(suffix, ptype = character(), size = 2L)
+  na_matches <- arg_match(na_matches, c("never", "na"))
+
+  lazy_query(
+    query_type = "join",
+    x = x,
+    y = y,
+    vars = vars,
+    type = type,
+    by = by,
+    suffix = suffix,
+    na_matches = na_matches,
+    last_op = "join"
   )
 }
 
+join_check_vars <- function(vars) {
+  if (!vctrs::vec_is_list(vars)) {
+    # TODO use `cli_abort()` after https://github.com/r-lib/rlang/issues/1386
+    # is fixed
+    abort("`vars` must be a list", .internal = TRUE)
+  }
+
+  if (!identical(names(vars), c("alias", "x", "y", "all_x", "all_y"))) {
+    # TODO use `cli_abort()` after https://github.com/r-lib/rlang/issues/1386
+    # is fixed
+    abort("`vars` must have fields `alias`, `x`, `y`, `all_x`, and `all_y`", .internal = TRUE)
+  }
+
+  n <- vctrs::vec_size(vars$alias)
+  vctrs::vec_assert(vars$alias, character())
```

Comment:
These all should to use `arg` and `call` to generate maximally informative errors. We might need to pass the call env through a few functions, but it's probably worth it.